### PR TITLE
Allow test dependency versions to override runtime dependencies

### PR DESCRIPTION
**Pull request by jroper**
IDEA and SBT treat dependency scopes fundamentally differently.  In IDEA, there is  a single classpath, and dependencies are either included or not included in it, depending on the current scope.  In SBT, each scope may "evict" dependencies out of the classpath.

For example, if I have dependency in the runtime scope on com.foo:bar:1.1, and then I have a test dependency that transitively depends on com.foo:bar:1.2, then in my runtime scope in SBT, I will have com.foo:bar:1.1, while in test, that will be evicted, and I'll have com.foo:bar:1.2.  For a comparison with other build tools, in Maven for example, the behaviour here is that the test transitive dependency will result in the runtime dependency being upgraded to 1.2, while maintaining its runtime scope.

The way that SBT IDEA currently handles this is that it puts both dependencies on the classpath in the order declared in the build file.  If the test scoped dependency appears later in the buid file, this will cause issues, because 1.1 will come before 1.2, and so it will be found first, so if the transitive test dependency actually uses new features from 1.2, you'll get linkage errors when you run the tests in IDEA, where you won't in SBT.

A naive approach to solving this (that I tried) is to just put all test dependencies before runtime dependencies.  However, this causes another issue, it means the classpath is now in a completely different order to what SBT uses.  In my specific case, another test dependency was bringing a dependency that had changed its groupId, and my runtime scope was depending on the newer version with a different groupId.  In SBT, both dependencies were on the classpath, but since the test dependency was declared later in the classpath, the correct, newer dependency was being used, so it wasn't a problem.  But with in IDEA it was.

So, I've implemented an approach that as closely as possible simulates what SBT does.  It detects when SBT would evict a runtime dependency and replace it with a test dependency, and in that case, moves just that dependency to be just before the runtime dependency in the classpath.  The result is that the classpath order is almost the same as what SBT uses, but not quite in the case of these "evicted" dependencies.  There are still some edge cases where this could cause problems, but I think they would be very rare, if they ever did happen, and I don't think it's possible to 100% emulate SBTs behaviour in IDEA since they have fundamentally different notions of the capabilities of scopes.

Also included in this pull request is a nicer format for the name of the dependencies.  Instead of being named:

com_foo_bar_bar_1.2.3_test

They are now named:

SBT: com.foo.bar:bar:1.2.3

This matches how IDEA handles Maven dependencies, and looks much cleaner.  The old semantics for the filenames are still used.

### CODE REVIEW

File: src/main/scala/org/sbtidea/IdeaModuleDescriptor.scala [hunk 1]
```diff
@@ -165,4 +165,28 @@ class IdeaModuleDescriptor(val imlDir: File, projectRoot: File, val project: Sub
...
   }
+
+  /**
+   * SBT allows different versions of the same library in different scopes, evicting runtime/compile dependencies
+   * out of their scopes if a newer one is found in test when compiling/running tests, while still maintaining
+   * the order as defined in the build configuration. IDEA doesn't support this eviction.  We fake it by
+   * detecting evictors, and moving them up in the classpath to just before the evictees.
+   */
+  def promoteTestEvictors(libraries: Seq[IdeaModuleLibRef]) = {
+    val evictions = libraries.groupBy(_.library.evictionId)
```
Comment (jozic) [inline src/main/scala/org/sbtidea/IdeaModuleDescriptor.scala]: Hey James, could this be written like this?

```
val evictions = for {
  (evictionId, libs) <- libraries.groupBy(_.library.evictionId) if libs.size > 1
  testLib <- libs.find(_.config == IdeaLibrary.TestScope)
} yield
  testLib -> libs.filterNot(_ == testLib)
```

**Comment by mpeltonen** [comment]
Looks ok to me at first glance. Need to test some more with different projects before merging though, to better ensure that this does not break anything in the common cases anyway.

If the for comprehension that @jozic proposed above is semantically equivalent to pull request filter/map/flatMap code, it would be much clearer.

**Comment by jroper** [comment]
The for comprehension is semantically equivalent, and very nice.  I've updated the pull request.

**Comment by mpeltonen** [comment]
Merged, thanks!

I made few small changes/fixes which I noticed when fixing the scripted tests after the merge (see commits for details):
- Used the name as the filename after all instead of id, as I noticed that Idea renames the files based on name anyway
- Naming for scala instance libraries was a bit inconsistent
- Configuration/scope must be appended to the name too, as we can have dependencies which differ only by scope/configuration and these need to have separate names

**Pull request merged**